### PR TITLE
Quick pick: improve eventing on value change

### LINKED ISSUE #13729: Spelling Mistake in error
- VSCode Version: Code 1.5.3 (5be4091987a98e3870d89d630eb87be6d9bafd27, 2016-09-22T08:50:42.185Z)
- OS Version: Windows_NT ia32 6.3.9600

Steps to Reproduce:

1. Debugging Nodejs
2. Hit a break point
3. Continue

The word should be "respond" instead of "repond".
capture

Its just a little typo, doesn't really matter but should be correct.

### LINKED ISSUE #137279: QuickPick: filter not reset on value set programmatically
Does this issue occur when all extensions are disabled?: Yes/No

- VS Code Version: 1.62.2
- OS Version: Darwin x64 20.6.0

Steps to Reproduce:

1. Setup a QuickPick as follows
```typescript
const qp = window.createQuickPick();
qp.items = [
	{ label: "This is first" },
	{ label: "This is second" },
];
const clearIcon = { iconPath: new ThemeIcon('trash') };
qp.buttons = [clearIcon];
const disposables = [
	qp.onDidTriggerButton((button) => {
		if (button === clearIcon) {
			qp.value = "";
		}
	}),
	qp
];
qp.onDidHide(() => {
	for (const d of disposables) {
		d.dispose();
	}
});
qp.show();
```

2. Run the code above
3. When the quickpick is shown, type `first`
4. The menu should only show one item because of the filter word `first`
5. Click the trash button to clear the value programmatically
6.

ACTUAL: The filter doesn't update, and only show one menu items
EXPECTED: The filter updates, and show two menu items

https://user-images.githubusercontent.com/3455662/141921763-0521887c-ef66-4280-8db4-c1b5cd7b2e17.mov

### PR DESCRIPTION
**Pull request by bpasero**
(from a 1on1 with Tyler): The intent of this change is to let the input box event handler be the only one that emits events on value changes, and to address the refiltering on programmatic value change (which fixes #137279).

### CODE REVIEW

**Comment by TylerLeonhardt** [comment]
Looking at this with slightly fresher eyes from this morning, this change breaks this:

```
                const quickpick = controller.createQuickPick();
		quickpick.onDidChangeValue((e) => value = e);
		quickpick.show();
		quickpick.value = 'changed';
```

where that first `quickpick.value` does not get a changed triggered. I feel as if we should be doing something based on the visibility of the quickpick?

File: src/vs/base/parts/quickinput/browser/quickInput.ts [hunk 1]
```diff
@@ -483,19 +484,25 @@ class QuickPick<T extends IQuickPickItem> extends QuickInput implements IQuickPi
 	}
 
 	get value() {
-		return this._value;
+		return this._value ?? '';
 	}
 
 	set value(value: string) {
 		if (this._value !== value) {
+			if (typeof this._value !== 'undefined') {
```
Comment (TylerLeonhardt) [inline src/vs/base/parts/quickinput/browser/quickInput.ts:492]: If I change this to `if (typeof this._value !== 'undefined' || this.visible) {` then all tests appear to pass... however, I need to get a better understanding I think.

**Comment by bpasero** [comment]
Very good catch 👍 . Let's wait for input from @chrmarti how this should work.

File: src/vs/base/parts/quickinput/browser/quickInput.ts (hunk 1 above)
Comment (chrmarti) [reply src/vs/base/parts/quickinput/browser/quickInput.ts:492]: The listener on the textbox below in `show()` is only registered while the QuickPick is visible. Calling `this.onDidChangeValueEmitter.fire(this._value)` here might still make sense when the QuickPick is hidden. Or is that specifically what you want to change?

Comment (bpasero) [reply src/vs/base/parts/quickinput/browser/quickInput.ts:492]: @chrmarti the issue is https://github.com/microsoft/vscode/issues/137279 where when you set the value of the input box programmatically, we are not ending up refiltering the picker, because of this logic:
* `this._value = value || '';` is being called [here](https://github.com/microsoft/vscode/blob/26fe37ca3d9dcc934ffcacd82443a38c9b192dd0/src/vs/base/parts/quickinput/browser/quickInput.ts#L491) to update the model of the picker
* this is followed by a call to `update`
* the input box updates its value [here](https://github.com/microsoft/vscode/blob/26fe37ca3d9dcc934ffcacd82443a38c9b192dd0/src/vs/base/parts/quickinput/browser/quickInput.ts#L969)
* the `onDidChange` listener fires from the input box
* however the picker refuses to refilter because it thinks the value has not changed:

https://github.com/microsoft/vscode/blob/26fe37ca3d9dcc934ffcacd82443a38c9b192dd0/src/vs/base/parts/quickinput/browser/quickInput.ts#L737-L739

Comment (chrmarti) [reply src/vs/base/parts/quickinput/browser/quickInput.ts:492]: Would Tyler's fix improve if it was wrapped in an `if (this.visible)`? (A similar approach would be to move Tyler's change into `update()` which also updates the filtering when the items update.)

On not checking `value === this.value`: We would want to avoid triggering additional events on the API, but we could filter unconditionally.

Comment (bpasero) [reply src/vs/base/parts/quickinput/browser/quickInput.ts:492]: @chrmarti but I think there really is a bug in your logic how you decide to refilter or not. 

This code ensures to only update the input box when the value differs:

https://github.com/microsoft/vscode/blob/26fe37ca3d9dcc934ffcacd82443a38c9b192dd0/src/vs/base/parts/quickinput/browser/quickInput.ts#L968-L970

But then in the change listener this condition checks on a different value:

https://github.com/microsoft/vscode/blob/26fe37ca3d9dcc934ffcacd82443a38c9b192dd0/src/vs/base/parts/quickinput/browser/quickInput.ts#L737-L739

In other words: we knew the value would change and updated the input box but then in the listener for the change we forget about this fact?

Comment (chrmarti) [reply src/vs/base/parts/quickinput/browser/quickInput.ts:492]: The way I think about it: The first propagates the value from the model object to the input box and forgets to filter (the bug you are fixing). The second propagates the value from the input box to the model object and correctly filters.

Refiltering in `update()` (your first code reference, maybe further down in that method after the items were updated) if the value changed seems like it might address this first case.

I haven't debugged into it yet, so I might be missing something.

Comment (bpasero) [reply src/vs/base/parts/quickinput/browser/quickInput.ts:492]: Yeah good point, I have opened https://github.com/microsoft/vscode/pull/137857 now with a different strategy that goes more into that direction.

**Comment by bpasero** [comment]
Continues in https://github.com/microsoft/vscode/pull/137857
